src/snarks/groth16/mpc/chacha_rng.cpp

```diff
@@ -0,0 +1,65 @@
+#include "snarks/groth16/mpc/chacha_rng.hpp"
+
+#include <algorithm>
+#include <cstring>
+#include <sodium/crypto_stream_chacha20.h>
+
+namespace libzeth
+{
+
+chacha_rng::chacha_rng(const void *seed, size_t seed_size)
+    : data_used(sizeof(data))
+{
+    seed_size = std::min(seed_size, sizeof(key));
+    memcpy(key, seed, seed_size);
+    if (seed_size < sizeof(key)) {
+        memset(((uint8_t *)key) + seed_size, 0, sizeof(key) - seed_size);
+    }
```

Comment:
Here, you follow the `from_seed` function  https://github.com/rust-random/rand/blob/master/rand_chacha/src/chacha.rs#L144, right?

Reply:
I've added comments here and in the other functions referring to the rust implementation.  Also renamed the internal `populate` method to match `update` in rust.

Reply:
Perfect, thanks